File: loan_manager/server/routers/repayments.py
from datetime import date, datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, case
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import RepaymentPlan, Order, Customer, User
from ..schemas import RepaymentPlanCreate, RepaymentPlanUpdate, RepaymentPlanOut
from ..auth_deps import get_active_user

router = APIRouter(prefix="/api/repayments", tags=["repayments"])

# Throttle: only run overdue sync once per minute per tenant
_last_sync: dict[int, datetime] = {}
_SYNC_INTERVAL_SECONDS = 60


def auto_sync_overdue(db: Session, tenant_id: int):
    """Mark all past-due unpaid plans as '逾期未还' and sync order statuses.
    Throttled to at most once per minute per tenant to avoid repeated full scans.
    """
    now = datetime.now()
    last = _last_sync.get(tenant_id)
    if last and (now - last).total_seconds() < _SYNC_INTERVAL_SECONDS:
        return
    _last_sync[tenant_id] = now

    today = date.today()
    overdue_plans = db.query(RepaymentPlan).filter(
        RepaymentPlan.tenant_id == tenant_id,
        RepaymentPlan.due_date < today,
        RepaymentPlan.status == "待还",
    ).all()
    if not overdue_plans:
        return
    affected_order_ids = set()
    for rp in overdue_plans:
        rp.status = "逾期未还"
        affected_order_ids.add(rp.order_id)
    db.flush()
    for oid in affected_order_ids:
        _sync_order_status(db, oid)
    db.commit()


_PAID_STATUSES = {"已还", "逾期还款"}


def _sync_order_status(db: Session, order_id: int):
    order = db.query(Order).filter(Order.id == order_id).first()
    if not order:
        return

    plans = db.query(RepaymentPlan).filter(RepaymentPlan.order_id == order_id).all()
    if not plans:
        return

    # All periods paid → mark order as fully settled
    if all(p.status in _PAID_STATUSES for p in plans):
        order.status = "已结清"
        return

    # If order was previously settled but a payment was undone, revert to active
    if order.status == "已结清":
        order.status = "已通过"

    # Overdue logic
    has_overdue = any(p.status == "逾期未还" for p in plans)
    if has_overdue and order.status != "逾期":
        order.status = "逾期"
    elif not has_overdue and order.status == "逾期":
        order.status = "已通过"


def _plan_to_dict(rp: RepaymentPlan, order: Order | None = None, customer_name: str = "") -> dict:
    """Convert a single RepaymentPlan ORM object to response dict."""
    d = RepaymentPlanOut.model_validate(rp).model_dump()
    d["order_no"] = order.order_no if order else ""
    d["customer_name"] = customer_name
    d["order_total_price"] = float((order.unit_price + order.processing_fee) * order.weight) if order else 0
    d["order_down_payment"] = float(order.down_payment) if order else 0
    d["credit_reported"] = order.credit_reported if order else False
    d["credit_reported_at"] = order.credit_reported_at.strftime("%Y-%m-%d %H:%M") if order and order.credit_reported_at else None
    d["lawsuit_filed"] = order.lawsuit_filed if order else False
    d["lawsuit_filed_at"] = order.lawsuit_filed_at.strftime("%Y-%m-%d %H:%M") if order and order.lawsuit_filed_at else None
    return d


# ── Summary endpoint (one row per order, used by the list view) ──────────────

@router.get("/summary")
def list_repayments_summary(
    keyword: str = Query(""),
    user: User = Depends(get_active_user),
    db: Session = Depends(get_db),
):
    """Return one aggregated row per order instead of all individual plans.
    Much faster for the list view — detail plans are loaded on demand.
    """
    auto_sync_overdue(db, user.tenant_id)

    order_q = db.query(Order).filter(
        Order.tenant_id == user.tenant_id,
        Order.installment_periods > 0,
    )
    if keyword:
        cids = [cid for (cid,) in db.query(Customer.id).filter(
            Customer.tenant_id == user.tenant_id,
            Customer.name.contains(keyword),
        ).all()]
        order_q = order_q.filter(
            Order.order_no.contains(keyword) | Order.customer_id.in_(cids)
        )
    orders = order_q.order_by(Order.id.desc()).all()
    if not orders:
        return []

    # Batch-load customers
    cids_set = {o.customer_id for o in orders if o.customer_id}
    cmap = {c.id: c.name for c in db.query(Customer).filter(Customer.id.in_(cids_set)).all()} if cids_set else {}

    # Aggregate stats per order in one SQL query
    order_ids = [o.id for o in orders]
    stats_rows = db.query(
        RepaymentPlan.order_id,
        func.count(RepaymentPlan.id).label("total_periods"),
        func.sum(RepaymentPlan.total_amount).label("installment_total"),
        func.sum(RepaymentPlan.paid_amount).label("paid_total"),
        func.sum(case((RepaymentPlan.status.in_(["已还", "逾期还款"]), 1), else_=0)).label("paid_count"),
    ).filter(
        RepaymentPlan.tenant_id == user.tenant_id,
        RepaymentPlan.order_id.in_(order_ids),
    ).group_by(RepaymentPlan.order_id).all()

    stats_map = {s.order_id: s for s in stats_rows}

    result = []
    for o in orders:
        s = stats_map.get(o.id)
        if not s:
            continue
        result.append({
            "order_id": o.id,
            "order_no": o.order_no,
            "customer_name": cmap.get(o.customer_id, ""),
            "order_total_price": float((o.unit_price + o.processing_fee) * o.weight),
            "credit_reported": o.credit_reported,
            "credit_reported_at": o.credit_reported_at.strftime("%Y-%m-%d %H:%M") if o.credit_reported_at else None,
            "lawsuit_filed": o.lawsuit_filed,
            "lawsuit_filed_at": o.lawsuit_filed_at.strftime("%Y-%m-%d %H:%M") if o.lawsuit_filed_at else None,
            "total_periods": s.total_periods,
            "installment_total": round(float(s.installment_total or 0), 2),
            "paid_total": round(float(s.paid_total or 0), 2),
            "paid_count": s.paid_count or 0,
        })
    return result


# ── Per-order detail (loaded on demand when user clicks 还款明细) ─────────────

@router.get("/")
def list_repayments(
    order_id: int | None = Query(None),
    keyword: str = Query(""),
    status: str = Query(""),
    skip: int = 0,
    user: User = Depends(get_active_user),
    db: Session = Depends(get_db),
):
    q = db.query(RepaymentPlan).filter(RepaymentPlan.tenant_id == user.tenant_id)
    if order_id is not None:
        q = q.filter(RepaymentPlan.order_id == order_id)
    if keyword:
        order_ids = (
            db.query(Order.id)
            .outerjoin(Customer, Order.customer_id == Customer.id)
            .filter(
                Order.tenant_id == user.tenant_id,
                Order.order_no.contains(keyword)
                | Customer.name.contains(keyword)
                | Customer.phone.contains(keyword)
            )
            .all()
        )
        ids = [oid for (oid,) in order_ids]
        q = q.filter(RepaymentPlan.order_id.in_(ids)) if ids else q.filter(False)
    if status:
        q = q.filter(RepaymentPlan.status == status)
    plans = q.order_by(RepaymentPlan.order_id, RepaymentPlan.period_no).offset(skip).all()

    # Batch-load orders and customers
    oids = {p.order_id for p in plans}
    order_map: dict[int, Order] = {}
    cmap: dict[int, str] = {}
    if oids:
        orders_list = db.query(Order).filter(Order.id.in_(oids)).all()
        order_map = {o.id: o for o in orders_list}
        cids = {o.customer_id for o in orders_list if o.customer_id}
        if cids:
            cmap = {c.id: c.name for c in db.query(Customer).filter(Customer.id.in_(cids)).all()}

    return [_plan_to_dict(p, order_map.get(p.order_id), cmap.get(order_map.get(p.order_id, None) and order_map[p.order_id].customer_id, "")) for p in plans]


@router.get("/count")
def count_repayments(order_id: int | None = None, status: str = "", user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    q = db.query(RepaymentPlan).filter(RepaymentPlan.tenant_id == user.tenant_id)
    if order_id is not None:
        q = q.filter(RepaymentPlan.order_id == order_id)
    if status:
        q = q.filter(RepaymentPlan.status == status)
    return {"count": q.count()}


@router.post("/", response_model=RepaymentPlanOut)
def create_repayment(data: RepaymentPlanCreate, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    rp = RepaymentPlan(**data.model_dump(), tenant_id=user.tenant_id)
    db.add(rp)
    db.flush()
    _sync_order_status(db, rp.order_id)
    db.commit()
    db.refresh(rp)
    order = db.query(Order).filter(Order.id == rp.order_id).first()
    customer_name = ""
    if order:
        c = db.query(Customer).filter(Customer.id == order.customer_id).first()
        customer_name = c.name if c else ""
    return _plan_to_dict(rp, order, customer_name)


@router.post("/batch", response_model=list[RepaymentPlanOut])
def batch_create_repayments(items: list[RepaymentPlanCreate], user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    results = []
    order_ids = set()
    for data in items:
        rp = RepaymentPlan(**data.model_dump(), tenant_id=user.tenant_id)
        db.add(rp)
        db.flush()
        results.append(rp)
        order_ids.add(rp.order_id)
    for oid in order_ids:
        _sync_order_status(db, oid)
    db.commit()
    for rp in results:
        db.refresh(rp)
    # Batch-load for response
    oids = {rp.order_id for rp in results}
    order_map = {o.id: o for o in db.query(Order).filter(Order.id.in_(oids)).all()}
    cids = {o.customer_id for o in order_map.values() if o.customer_id}
    cmap = {c.id: c.name for c in db.query(Customer).filter(Customer.id.in_(cids)).all()} if cids else {}
    return [_plan_to_dict(rp, order_map.get(rp.order_id), cmap.get(getattr(order_map.get(rp.order_id), 'customer_id', None), "")) for rp in results]


@router.put("/{plan_id}", response_model=RepaymentPlanOut)
def update_repayment(plan_id: int, data: RepaymentPlanUpdate, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    rp = db.query(RepaymentPlan).filter(RepaymentPlan.id == plan_id, RepaymentPlan.tenant_id == user.tenant_id).first()
    if not rp:
        raise HTTPException(status_code=404, detail="还款记录不存在")
    for k, v in data.model_dump(exclude_none=True).items():
        setattr(rp, k, v)
    db.flush()
    _sync_order_status(db, rp.order_id)
    db.commit()
    db.refresh(rp)
    order = db.query(Order).filter(Order.id == rp.order_id).first()
    customer_name = ""
    if order:
        c = db.query(Customer).filter(Customer.id == order.customer_id).first()
        customer_name = c.name if c else ""
    return _plan_to_dict(rp, order, customer_name)


@router.delete("/by-order/{order_id}")
def delete_repayments_by_order(order_id: int, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    """Delete all repayment plans for a given order (bulk delete)."""
    db.query(RepaymentPlan).filter(
        RepaymentPlan.order_id == order_id,
        RepaymentPlan.tenant_id == user.tenant_id,
    ).delete(synchronize_session=False)
    db.commit()
    return {"message": "删除成功"}


@router.delete("/{plan_id}")
def delete_repayment(plan_id: int, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    rp = db.query(RepaymentPlan).filter(RepaymentPlan.id == plan_id, RepaymentPlan.tenant_id == user.tenant_id).first()
    if not rp:
        raise HTTPException(status_code=404, detail="还款记录不存在")
    db.delete(rp)
    db.commit()
    return {"message": "删除成功"}

File: loan_manager/loan-manager-cf/src/routes/repayments.ts
import { Hono } from 'hono';
import { drizzle } from 'drizzle-orm/d1';
import { eq, and, like, or, sql, desc, asc, inArray } from 'drizzle-orm';
import { repaymentPlans, orders, customers, warehouseEntries } from '../db/schema';
import { authMiddleware, activeUserMiddleware, writeProtectMiddleware } from '../middleware/auth';
import type { Env } from '../lib/types';

const repaymentsRoute = new Hono<Env>();
repaymentsRoute.use('*', authMiddleware, activeUserMiddleware, writeProtectMiddleware);

const PAID_STATUSES = new Set(['已还', '逾期还款']);

function todayStr() { return new Date().toISOString().slice(0, 10); }

export async function autoSyncOverdue(db: ReturnType<typeof drizzle>, tenantId: number) {
  const today = todayStr();
  // Bulk update all overdue plans in one query
  const result = await db.update(repaymentPlans)
    .set({ status: '逾期未还' })
    .where(and(
      eq(repaymentPlans.tenantId, tenantId),
      sql`${repaymentPlans.dueDate} < ${today}`,
      eq(repaymentPlans.status, '待还'),
    ))
    .returning({ orderId: repaymentPlans.orderId });

  if (!result.length) return;

  const affectedOrderIds = [...new Set(result.map(r => r.orderId))];
  await batchSyncOrderStatus(db, affectedOrderIds);
}

async function syncOrderStatus(db: ReturnType<typeof drizzle>, orderId: number) {
  const orderIds = [orderId];
  await batchSyncOrderStatus(db, orderIds);
}

async function batchSyncOrderStatus(db: ReturnType<typeof drizzle>, orderIds: number[]) {
  if (!orderIds.length) return;

  const orderList = await db.select().from(orders).where(inArray(orders.id, orderIds));
  if (!orderList.length) return;

  const allPlans = await db.select({
    orderId: repaymentPlans.orderId,
    status: repaymentPlans.status,
  }).from(repaymentPlans).where(inArray(repaymentPlans.orderId, orderIds));

  // Group plans by order
  const plansByOrder: Record<number, string[]> = {};
  for (const p of allPlans) {
    if (!plansByOrder[p.orderId]) plansByOrder[p.orderId] = [];
    plansByOrder[p.orderId].push(p.status || '');
  }

  for (const order of orderList) {
    const statuses = plansByOrder[order.id];
    if (!statuses || !statuses.length) continue;

    const allPaid = statuses.every(s => PAID_STATUSES.has(s));
    if (allPaid) {
      if (order.status !== '已结清') await db.update(orders).set({ status: '已结清' }).where(eq(orders.id, order.id));
      continue;
    }

    if (order.status === '已结清') {
      await db.update(orders).set({ status: '已通过' }).where(eq(orders.id, order.id));
    }

    const hasOverdue = statuses.some(s => s === '逾期未还');
    if (hasOverdue && order.status !== '逾期') {
      await db.update(orders).set({ status: '逾期' }).where(eq(orders.id, order.id));
    } else if (!hasOverdue && order.status === '逾期') {
      await db.update(orders).set({ status: '已通过' }).where(eq(orders.id, order.id));
    }
  }
}

function planOut(p: any, order: any = null, customerName = '') {
  return {
    id: p.id, tenant_id: p.tenantId, order_id: p.orderId,
    period_no: p.periodNo, due_date: p.dueDate,
    principal: p.principal || '0.00', interest: p.interest || '0.00',
    total_amount: p.totalAmount || '0.00', paid_amount: p.paidAmount || '0.00',
    paid_date: p.paidDate || null, payment_account: p.paymentAccount || '',
    status: p.status || '待还',
    order_no: order?.orderNo || '',
    customer_name: customerName,
    order_total_price: order ? (parseFloat(order.unitPrice || '0') + parseFloat(order.processingFee || '0')) * parseFloat(order.weight || '0') : 0,
    order_down_payment: order ? parseFloat(order.downPayment || '0') : 0,
    credit_reported: order?.creditReported || false,
    credit_reported_at: order?.creditReportedAt || null,
    lawsuit_filed: order?.lawsuitFiled || false,
    lawsuit_filed_at: order?.lawsuitFiledAt || null,
  };
}

// ── Summary (one row per order) ──────────────────────────
repaymentsRoute.get('/summary', async (c) => {
  const user = c.get('user');
  const keyword = c.req.query('keyword') || '';
  const db = drizzle(c.env.DB);
  const tid = user.tenantId!;

  // autoSyncOverdue moved to cron job for performance
  const conditions: any[] = [eq(orders.tenantId, tid), sql`${orders.installmentPeriods} > 0`];
  if (keyword) {
    const custRows = await db.select({ id: customers.id }).from(customers)
      .where(and(eq(customers.tenantId, tid), like(customers.name, `%${keyword}%`)));
    const cids = custRows.map(r => r.id);
    if (cids.length > 0) {
      conditions.push(or(like(orders.orderNo, `%${keyword}%`), inArray(orders.customerId, cids)));
    } else {
      conditions.push(like(orders.orderNo, `%${keyword}%`));
    }
  }

  const orderList = await db.select().from(orders).where(and(...conditions)).orderBy(asc(orders.orderDate), asc(orders.id));
  if (!orderList.length) return c.json([]);

  // Batch load customer info
  const cids = [...new Set(orderList.map(o => o.customerId))];
  const cmap: Record<number, { name: string; phone: string; idCard: string; address: string }> = {};
  if (cids.length > 0) {
    const custs = await db.select().from(customers).where(inArray(customers.id, cids));
    custs.forEach(c => { cmap[c.id] = { name: c.name, phone: c.phone || '', idCard: c.idCard || '', address: c.address || '' }; });
  }

  // Aggregate per order using SQL (much faster than loading all rows)
  const orderIds = orderList.map(o => o.id);
  const stats = await db.select({
    orderId: repaymentPlans.orderId,
    totalPeriods: sql<number>`count(*)`,
    installmentTotal: sql<number>`ROUND(SUM(CAST(${repaymentPlans.totalAmount} AS REAL)), 2)`,
    paidTotal: sql<number>`ROUND(SUM(CAST(${repaymentPlans.paidAmount} AS REAL)), 2)`,
    paidCount: sql<number>`SUM(CASE WHEN ${repaymentPlans.status} IN ('已还','逾期还款') THEN 1 ELSE 0 END)`,
  }).from(repaymentPlans)
    .where(and(eq(repaymentPlans.tenantId, tid), inArray(repaymentPlans.orderId, orderIds)))
    .groupBy(repaymentPlans.orderId);

  const statsMap: Record<number, typeof stats[0]> = {};
  for (const s of stats) statsMap[s.orderId] = s;

  // Batch load warehouse costs by order_no (stored in notes field)
  const orderNos = orderList.map(o => o.orderNo).filter(Boolean);
  const costMap: Record<string, number> = {};
  if (orderNos.length > 0) {
    const whCosts = await db.select({
      orderNo: warehouseEntries.notes,
      totalCost: sql<number>`ROUND(SUM(CAST(${warehouseEntries.totalPrice} AS REAL)), 2)`,
    }).from(warehouseEntries)
      .where(and(eq(warehouseEntries.tenantId, tid), inArray(warehouseEntries.notes, orderNos)))
      .groupBy(warehouseEntries.notes);
    for (const wh of whCosts) {
      if (wh.orderNo) costMap[wh.orderNo] = wh.totalCost || 0;
    }
  }

  const result = orderList.filter(o => statsMap[o.id]).map(o => {
    const s = statsMap[o.id];
    const orderTotal = (parseFloat(o.unitPrice || '0') + parseFloat(o.processingFee || '0')) * parseFloat(o.weight || '0');
    const downPayment = parseFloat(o.downPayment || '0');
    const installTotal = s.installmentTotal || 0;
    const cost = costMap[o.orderNo] || 0;
    const managerComm = Math.round(orderTotal * 0.02 * 100) / 100;
    const operatorComm = Math.round(orderTotal * 0.01 * 100) / 100;
    const paidTotal = downPayment + (s.paidTotal || 0);
    const profit = Math.round((paidTotal - cost - managerComm - operatorComm) * 100) / 100;
    const cust = cmap[o.customerId] || { name: '', phone: '', idCard: '', address: '' };
    return {
      order_id: o.id, order_no: o.orderNo, order_date: o.orderDate || '',
      customer_name: cust.name,
      customer_phone: cust.phone,
      customer_id_card: cust.idCard,
      customer_address: cust.address,
      order_total_price: orderTotal,
      down_payment: downPayment,
      cost,
      manager_commission: managerComm,
      operator_commission: operatorComm,
      profit,
      profit_rate: orderTotal > 0 ? Math.round(profit / orderTotal * 10000) / 100 : 0,
      credit_reported: o.creditReported || false,
      credit_reported_at: o.creditReportedAt || null,
      lawsuit_filed: o.lawsuitFiled || false,
      lawsuit_filed_at: o.lawsuitFiledAt || null,
      total_periods: s.totalPeriods,
      installment_total: installTotal,
      paid_total: s.paidTotal || 0,
      paid_count: s.paidCount || 0,
    };
  });
  return c.json(result);
});

// ── List (per-order detail) ──────────────────────────────
repaymentsRoute.get('/', async (c) => {
  const user = c.get('user');
  const orderId = c.req.query('order_id');
  const keyword = c.req.query('keyword') || '';
  const status = c.req.query('status') || '';
  const skip = parseInt(c.req.query('skip') || '0');
  const db = drizzle(c.env.DB);
  const tid = user.tenantId!;

  const conditions: any[] = [eq(repaymentPlans.tenantId, tid)];
  if (orderId) conditions.push(eq(repaymentPlans.orderId, parseInt(orderId)));
  if (status) conditions.push(eq(repaymentPlans.status, status));
  if (keyword) {
    // Search by order_no or customer name
    const matchOrders = await db.select({ id: orders.id }).from(orders)
      .leftJoin(customers, eq(orders.customerId, customers.id))
      .where(and(
        eq(orders.tenantId, tid),
        or(like(orders.orderNo, `%${keyword}%`), like(customers.name, `%${keyword}%`), like(customers.phone, `%${keyword}%`)),
      ));
    const oids = matchOrders.map(r => r.id);
    if (oids.length > 0) {
      conditions.push(inArray(repaymentPlans.orderId, oids));
    } else {
      return c.json([]);
    }
  }

  const plans = await db.select().from(repaymentPlans).where(and(...conditions))
    .orderBy(asc(repaymentPlans.orderId), asc(repaymentPlans.periodNo)).offset(skip);

  // Batch load orders and customers
  const oids = [...new Set(plans.map(p => p.orderId))];
  const orderMap: Record<number, any> = {};
  const cmap: Record<number, string> = {};
  if (oids.length > 0) {
    const orderList = await db.select().from(orders).where(inArray(orders.id, oids));
    orderList.forEach(o => { orderMap[o.id] = o; });
    const cids = [...new Set(orderList.map(o => o.customerId))];
    if (cids.length > 0) {
      const custs = await db.select().from(customers).where(inArray(customers.id, cids));
      custs.forEach(c => { cmap[c.id] = c.name; });
    }
  }

  return c.json(plans.map(p => {
    const order = orderMap[p.orderId];
    return planOut(p, order, cmap[order?.customerId] || '');
  }));
});

// ── Count ────────────────────────────────────────────────
repaymentsRoute.get('/count', async (c) => {
  const user = c.get('user');
  const orderId = c.req.query('order_id');
  const status = c.req.query('status') || '';
  const db = drizzle(c.env.DB);

  const conditions: any[] = [eq(repaymentPlans.tenantId, user.tenantId!)];
  if (orderId) conditions.push(eq(repaymentPlans.orderId, parseInt(orderId)));
  if (status) conditions.push(eq(repaymentPlans.status, status));

  const result = await db.select({ count: sql<number>`count(*)` }).from(repaymentPlans).where(and(...conditions));
  return c.json({ count: result[0]?.count || 0 });
});

// ── Create ───────────────────────────────────────────────
repaymentsRoute.post('/', async (c) => {
  const user = c.get('user');
  const body = await c.req.json();
  const db = drizzle(c.env.DB);

  const result = await db.insert(repaymentPlans).values({
    tenantId: user.tenantId,
    orderId: body.order_id,
    periodNo: body.period_no,
    dueDate: body.due_date,
    principal: body.principal || '0.00',
    interest: body.interest || '0.00',
    totalAmount: body.total_amount || '0.00',
    paidAmount: body.paid_amount || '0.00',
    paidDate: body.paid_date || null,
    paymentAccount: body.payment_account || '',
    status: body.status || '待还',
  }).returning();
  const rp = result[0];

  await syncOrderStatus(db, rp.orderId);

  const order = await db.select().from(orders).where(eq(orders.id, rp.orderId)).get();
  let customerName = '';
  if (order) {
    const cust = await db.select().from(customers).where(eq(customers.id, order.customerId)).get();
    customerName = cust?.name || '';
  }
  return c.json(planOut(rp, order, customerName));
});

// ── Batch create ─────────────────────────────────────────
repaymentsRoute.post('/batch', async (c) => {
  const user = c.get('user');
  const items: any[] = await c.req.json();
  const db = drizzle(c.env.DB);
  const results: any[] = [];
  const orderIds = new Set<number>();

  for (const body of items) {
    const result = await db.insert(repaymentPlans).values({
      tenantId: user.tenantId,
      orderId: body.order_id,
      periodNo: body.period_no,
      dueDate: body.due_date,
      principal: body.principal || '0.00',
      interest: body.interest || '0.00',
      totalAmount: body.total_amount || '0.00',
      paidAmount: body.paid_amount || '0.00',
      paidDate: body.paid_date || null,
      paymentAccount: body.payment_account || '',
      status: body.status || '待还',
    }).returning();
    results.push(result[0]);
    orderIds.add(body.order_id);
  }

  for (const oid of orderIds) {
    await syncOrderStatus(db, oid);
  }

  // Batch load for response
  const oids = [...orderIds];
  const orderMap: Record<number, any> = {};
  const cmap: Record<number, string> = {};
  if (oids.length > 0) {
    const orderList = await db.select().from(orders).where(inArray(orders.id, oids));
    orderList.forEach(o => { orderMap[o.id] = o; });
    const cids = [...new Set(orderList.map(o => o.customerId))];
    if (cids.length > 0) {
      const custs = await db.select().from(customers).where(inArray(customers.id, cids));
      custs.forEach(c => { cmap[c.id] = c.name; });
    }
  }

  return c.json(results.map(rp => {
    const order = orderMap[rp.orderId];
    return planOut(rp, order, cmap[order?.customerId] || '');
  }));
});

// ── Batch update ────────────────────────────────────────
repaymentsRoute.put('/batch', async (c) => {
  const user = c.get('user');
  const items: any[] = await c.req.json();
  const db = drizzle(c.env.DB);
  const tid = user.tenantId!;

  // Load all plans in one query instead of N individual SELECTs
  const ids = items.map(i => i.id).filter(Boolean);
  const existing = await db.select({ id: repaymentPlans.id, orderId: repaymentPlans.orderId })
    .from(repaymentPlans)
    .where(and(eq(repaymentPlans.tenantId, tid), inArray(repaymentPlans.id, ids)));
  const existMap = new Map(existing.map(r => [r.id, r.orderId]));

  const orderIds = new Set<number>();
  for (const item of items) {
    const orderId = existMap.get(item.id);
    if (orderId === undefined) continue;
    await db.update(repaymentPlans).set({
      paidAmount: String(item.paid_amount ?? '0.00'),
      paidDate: item.paid_date || null,
      paymentAccount: item.payment_account || '',
      status: item.status || '待还',
    }).where(and(eq(repaymentPlans.id, item.id), eq(repaymentPlans.tenantId, tid)));
    orderIds.add(orderId);
  }

  await batchSyncOrderStatus(db, [...orderIds]);
  return c.json({ message: '保存成功', count: items.length });
});

// ── Update ───────────────────────────────────────────────
repaymentsRoute.put('/:id', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  const body = await c.req.json();
  const db = drizzle(c.env.DB);

  const rp = await db.select().from(repaymentPlans)
    .where(and(eq(repaymentPlans.id, id), eq(repaymentPlans.tenantId, user.tenantId!))).get();
  if (!rp) return c.json({ detail: '还款记录不存在' }, 404);

  const updates: Record<string, any> = {};
  if (body.period_no !== undefined) updates.periodNo = body.period_no;
  if (body.due_date !== undefined) updates.dueDate = body.due_date;
  if (body.principal !== undefined) updates.principal = body.principal;
  if (body.interest !== undefined) updates.interest = body.interest;
  if (body.total_amount !== undefined) updates.totalAmount = body.total_amount;
  if (body.paid_amount !== undefined) updates.paidAmount = body.paid_amount;
  if (body.paid_date !== undefined) updates.paidDate = body.paid_date;
  if (body.payment_account !== undefined) updates.paymentAccount = body.payment_account;
  if (body.status !== undefined) updates.status = body.status;

  await db.update(repaymentPlans).set(updates).where(eq(repaymentPlans.id, id));
  await syncOrderStatus(db, rp.orderId);

  const updated = await db.select().from(repaymentPlans).where(eq(repaymentPlans.id, id)).get();
  const order = await db.select().from(orders).where(eq(orders.id, rp.orderId)).get();
  let customerName = '';
  if (order) {
    const cust = await db.select().from(customers).where(eq(customers.id, order.customerId)).get();
    customerName = cust?.name || '';
  }
  return c.json(planOut(updated, order, customerName));
});

// ── Delete by order ──────────────────────────────────────
repaymentsRoute.delete('/by-order/:orderId', async (c) => {
  const user = c.get('user');
  const orderId = parseInt(c.req.param('orderId'));
  const db = drizzle(c.env.DB);

  await db.delete(repaymentPlans).where(and(
    eq(repaymentPlans.orderId, orderId),
    eq(repaymentPlans.tenantId, user.tenantId!),
  ));
  return c.json({ message: '删除成功' });
});

// ── Delete single ────────────────────────────────────────
repaymentsRoute.delete('/:id', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  const db = drizzle(c.env.DB);

  const rp = await db.select().from(repaymentPlans)
    .where(and(eq(repaymentPlans.id, id), eq(repaymentPlans.tenantId, user.tenantId!))).get();
  if (!rp) return c.json({ detail: '还款记录不存在' }, 404);

  await db.delete(repaymentPlans).where(eq(repaymentPlans.id, id));
  return c.json({ message: '删除成功' });
});

export default repaymentsRoute;
